Docs Update device memory pages (#3695)
ROCm-Developer-Tools/HIP: docs/how-to/hip_runtime_api/memory_management/device_memory.rst

.. meta::
  :description: This chapter describes the device memory of the HIP ecosystem
                ROCm software.
  :keywords: AMD, ROCm, HIP, GPU, device memory, global, constant, texture, surface, shared

.. _device_memory:

********************************************************************************
Device memory
********************************************************************************

Device memory is random access memory that is physically located on a GPU. In
general it is memory with a bandwidth that is an order of magnitude higher
compared to RAM available to the host. That high bandwidth is only available to
on-device accesses, accesses from the host or other devices have to go over a

> special interface, usually the PCIe bus or AMD Infinity Fabric, which is

Reviewer: ```suggestion
special interface which is considerably slower, usually the PCIe bus or AMD Infinity Fabric.
```

> considerably slower.

Reviewer: ```suggestion
```

On certain architectures like APUs, the GPU and CPU share the same physical
memory.

There is also a special local data share on-chip directly accessible to the
:ref:`compute units <hardware_implementation>`, that can be used for shared
memory.

The physical device memory can be used to back up several different memory

> spaces in HIP:

Reviewer: ```suggestion
spaces in HIP, as described below. 
```

Global memory is the general read-write accessible memory visible to all threads
on a given device. Since variables located in global memory have to be marked
with the ``__device__`` qualifier, this memory space is also referred to as
device memory.

Without explicitly copying it, it can only be accessed by the threads within a
kernel operating on the device, however :ref:`unified memory` can be used to
let the runtime manage this, if desired.

Allocating global memory
--------------------------------------------------------------------------------

This memory needs to be explicitly allocated.

It can be allocated from the host via the :ref:`HIP runtime memory management
functions <memory_management_reference>` like :cpp:func:`hipMalloc`, or can be
defined using the ``__device__`` qualifier on variables.

It can also be allocated within a kernel using ``malloc`` or ``new``.
The specified amount of memory is allocated by each thread that executes the
instructions. The recommended way to allocate the memory depends on the use
case. If the memory is intended to be shared between the threads of a block, it

> is generally beneficial to allocate on large block of memory, due the way the

Reviewer 2: due to the way...

Author: Fixed

memory is accessed.

.. note::
   Memory allocated within a kernel can only be freed in kernels, not by the HIP
   runtime on the host, like :cpp:func:`hipFree`. It is also not possible to

>    free device memory allocated on the host, with e.g. :cpp:func:`hipMalloc`, in

Reviewer: ```suggestion
   free device memory allocated on the host, with :cpp:func:`hipMalloc` for example, in
```

   example, in a kernel.

An example for how to share memory allocated within a kernel by only one thread
is given in the following example. In case the device memory is only needed for
communication between the threads in a single block, :ref:`shared_memory` is the
better option, but is also limited in size.

.. code-block:: cpp

  __global__ void kernel_memory_allocation(TYPE* pointer){
  // The pointer is stored in shared memory, so that all
  // threads of the block can access the pointer
  __shared__ int *memory;

  size_t blockSize = blockDim.x;
  constexpr size_t elementsPerThread = 1024;
  if(threadIdx.x == 0){
    // allocate memory in one contiguous block
    memory = new int[blockDim.x * elementsPerThread];
  }
  __syncthreads();

  // load pointer into thread-local variable to avoid
  // unnecessary accesses to shared memory
  int *localPtr = memory;

  // work with allocated memory, e.g. initialization
  for(int i = 0; i < elementsPerThread; ++i){
    // access in a contiguous way

>     localPtr[i * blockSize + threadIdx.x] = i;

Reviewer 3: Not sure of the purpose of this logics.
Anyway,  kernel will crash if blockSize -threadIdx.x <1

APIs, and provides performance benefits when accessing memory in a pattern where
the addresses are close to each other in a 2D or 3D representation of the
memory. It also provides additional features like filtering and addressing for
out-of-bounds accesses, which are further explained in :ref:`texture_fetching`.

The original use of the texture cache was also to take pressure off the global
memory and other caches, however on modern GPUs, that support textures, the L1
cache and texture cache are combined, so the main purpose is to make use of the
texture specific features.

To find out whether textures are supported on a device, query
:cpp:enumerator:`hipDeviceAttributeImageSupport`.

Using texture memory
--------------------------------------------------------------------------------

Textures are more complex than just a region of memory, so their layout has to
be specified. They are represented by ``hipTextureObject_t`` and created using
:cpp:func:`hipCreateTextureObject`.

The underlying memory is a 1D, 2D or 3D ``hipArray_t``, that needs to be
allocated using :cpp:func:`hipMallocArray`.

On the device side, texture objects are accessed using the ``tex1D/2D/3D``
functions.

> The texture management functions can be found in the :ref:`texture management

Reviewer: ```suggestion
The texture management functions can be found in the :ref:`Texture management
```

API reference <texture_management_reference>`

A full example for how to use textures can be found in the `ROCm texture
management example <https://github.com/ROCm/rocm-examples/blob/develop/HIP-Basic/texture_management/main.hip>`_

Surface memory
================================================================================

A read-write version of texture memory. It is created in the same way as a
texture, but with :cpp:func:`hipCreateSurfaceObject`.

Since surfaces also cached in the read-only texture cache, the changes written
back to the surface can't be observed in the same kernel. A new kernel has to be
launched in order to see the updated surface.

> The corresponding functions are listed in the :ref:`surface object API reference

Reviewer: ```suggestion
The corresponding functions are listed in the :ref:`Surface object API reference
```

variables between the threads within a block, but can also be used as scratch
pad memory.

Shared memory is not backed by the same physical memory as the other address
spaces. It is on-chip memory local to the :ref:`compute units
<hardware_implementation>`, providing low-latency, high-bandwidth access,
comparable to the L1 cache. It is however limited in size, and as it is
allocated per block, can restrict how many blocks can be scheduled to a compute
unit concurrently, thereby potentially reducing occupancy.

An overview of the size of the local data share (LDS), that backs up shared
memory, is given in the
:doc:`GPU hardware specifications <rocm:reference/gpu-arch-specs>`.

Allocate shared memory
--------------------------------------------------------------------------------

Memory can be dynamically allocated by declaring an ``extern __shared__`` array,
whose size can be set during kernel launch, which can then be accessed in the
kernel.

.. code-block:: cpp

  extern __shared__ int dynamic_shared[];
  __global__ void kernel(int sizeX, int sizeY, int totalSize){
    int* array1 = dynamic_shared;
    // array1 is interpreted as 2D of size:
    int array1_size = sizeX * sizeY;

>     int* array2 = &(array[sizeX * sizeY]);

Reviewer 2: Shoud it be the following?
int* array2 = &(array1[sizeX * sizeY]);

Author: Yep, thanks. I updated the whole snippet to be a bit more expressive. Also ran it through the compiler this time to verify it works :)
